synapse/storage/databases/main/events.py

```diff
@@ -1990,6 +1990,15 @@ def _set_push_actions_for_event_and_users_txn(
                 events_and_context.
         """
 
+        # Only non outlier events will have push actions associated with them,
+        # so let's filter them out. (This makes joining large rooms faster, as
+        # these queries took seconds to process all the state events).
```

Comment:
Let's see if I can reason through this.

* An event in the DAG is an outlier if we don't have the state at that point in the DAG
* "Don't have the state" meaning "hat state could always be computed if needed, but we don't have easy access to it"?

> Only non outlier events will have push actions associated with them

Not sure why this follows. What's a push action?

Reply:
Push actions is the output of running a user's push rules on an event and state and basically says if and how the user should be notified about the event (e.g. should they be bushed? Should it be highlighted? etc).

An `outlier` really means that its not part of the "timeline", i.e. the connected chunk(s) of the rooms graph that makes up the history/timeline we send down to the clients. Think timeline events as being a sequence of messages, whereas outliers being state that was set (e.g. room name) before you joined the room. You're right that we track state for events in the timeline and not for outliers, but the reason we won't have push actions for outliers is that we only calculate them for timeline events, as we only notify for events that have just been added to the timeline rather than for old state events.